### ISSUE TITLE
Build failure: SConf fails test for SDL.h [macro ordering problem]

### ISSUE THREAD

**Issue by Ambaire** [dxx-rebirth/dxx-rebirth#393]
### Environment
https://github.com/dxx-rebirth/dxx-rebirth/commit/9dafe85bbfcaf79f1577a28dc2b983bc210ac424

#### Operating System Environment
Microsoft Windows 8.1

#### CPU environment
* [ ] x86\_64 (64-bit Intel/AMD)

### Description

I followed the instructions at https://forum.dxx-rebirth.com/showthread.php?tid=857 and attempted to compile the program per the instructions there. I initially received the following error:

scons: *** Header SDL.h exists, but cannot compile an empty program. See sconf.log for details. Stop.

Full console output here, sorry, no text output. https://i.imgur.com/9Mw5IOO.png And sconf.log: https://pastebin.com/5Z62XXQc

I went on the r/Descent discord, and found someone who suggested changing line 842 of SConstruct from "" to ";Return 0;". It actually started compiling this time, but then got this error:

DXX-Archive.3: checking for C++11 inherited constructors with good unique_ptr<T[]> support...no
DXX-Archive.3: checking for C++11 variadic templates on constructors...no
scons: *** C++ compiler does not support constructor forwarding.  See sconf.log for details.  Stop.

#### Regression status

Unknown

### Steps to Reproduce

SConstruct line 842 changed from "" to ";return 0;"

Command executed: Per the forum directions above: scons.py CXX=/mingw/mingw32/bin/i686-w64-mingw32-g++ PKG_CONFIG=/mingw/mingw32/bin/pkgconf.exe RC=/mingw/mingw32/bin/windres host_platform=win32 -j4 > output.log 2>&!

output.log may be found here: https://pastebin.com/H1bHxQP2

sconf.log: https://pastebin.com/QtaCfnWh

dxxsconf.h: https://pastebin.com/Qx6Q5YTP

By the way, I tried registering for the DXX Rebirth forums, but it appears that the Recaptcha used is no longer valid and needs to be updated.

**Comment by vLKp** [comment]
@Ambaire: thank you for the report.  I cannot reproduce the failure on Linux, but I see at least part of your problem and can reproduce the problem when targetting Windows.  The fix suggested to you by the Reddit user was incorrect.  Implicit `return 0` is a quirk unique to `main` in C++ and is legal there, but only there.  It is not allowed for other functions.  [Add unit tests for valptridx parameter validation](/dxx-rebirth/dxx-rebirth/commit/2a82207e91f038ec20c8febcc4303657fb0d8389) reordered the parts of this test in a way that breaks the test when targeting Windows, but still works correctly on Linux.  SDL ships with a macro that remaps `main` to `SDL_main`; this macro is incorrect in the test environment and needs to be undefined.  On Linux, this macro comes in via the command-line's `pkg-config` flags and can be undefined by a `#undef` anywhere before the first use of `main`.  On Windows, it comes in on the command-line and then **again** through an SDL header.  The header has a platform-specific preprocessor guard, so Linux builds do not include this second definition.  Therefore, for Windows, the `#undef` must occur after including SDL headers.  Prior to that commit, it was.  Now, it is earlier and is ineffective.

[Fix SConf tests for Windows target](/dxx-rebirth/dxx-rebirth/commit/7d4c521382e25c9f3386235f3f3484ed2327c7e1) should resolve this for you.  Please sync and retry.

@Mako88: could you help @Ambaire with the reCAPTCHA issue?  @Ambaire: you may need to e-mail @Mako88.  He is not often active on Github these days.

**Comment by Ambaire** [comment]
Thanks for the reply. Full disclosure: I haven't done anything programming-related for about 6 years now, so I just followed the directions. Not sure how to sync what I have locally; I renamed the directory and recloned the /dxx-rebirth folder. 

When I attempted to recompile, I directed the output into a new file just to have it available.. looks like it failed, again. Hopefully the entire log will point out anything I did wrong. When I downloaded the compiler files per the forum thread, I used the latest versions unless a specific version was mentioned or there was only one version available.

Output log, using the 2>&1 option, seems to have caught some errors. Zip folder attached with output2.log, sconf.log, and dxxsconf.h.
[dxx-rebirth_compile_logs.zip](https://github.com/dxx-rebirth/dxx-rebirth/files/2152195/dxx-rebirth_compile_logs.zip)

**Comment by vLKp** [comment]
> I haven't done anything programming-related for about 6 years now, so I just followed the directions

That is fine.  If the directions are not sufficient for you to build this unaided, then the directions need improvement.  They are not supposed to require any domain expertise.

> Not sure how to sync what I have locally

`git pull`, as run from the root of your checkout, should work if you have not modified any version-controlled files.

Note that although I've read that forum thread, I've never actually executed it.  I can't say whether it worked when it was written (though I assume it did) or if it still works now.  I build always from Linux, even when targetting Windows.

As for your immediate problem - this expression is only used on >=gcc-6; a different construct is used in <=gcc-5.  That expression is accepted in gcc-7, but I can reproduce the failure even on Linux in gcc-6.  I normally use only the latest stable compiler, so I missed this until you reported it.

[Preprocess out glmprintf in normal builds](/dxx-rebirth/dxx-rebirth/commit/da268c6d21c6ce147d41cead3e2583279c889957) changes the macro to expand to nothing, which fixes the problem for me on `x86_64-pc-linux-gnu-g++-6.4.0`.  With that and some other fixes I just pushed, I can now build to completion using mingw-gcc-5.  I do not have mingw-gcc-6 installed at the moment, so I cannot be sure your next attempt will succeed.

Please pull and try again.

**Comment by Ambaire** [comment]
Thank you! It worked! I did the Rebirth install normally then copied the exe over and launched the game and everything seems to be working normally.

The instructions said to get gcc version 6.4.0 as "gcc 7 has a bug that results in a segfault when compiling Rebirth for Windows. I have reported it upstream.)". Is that no longer the case?

**Comment by vLKp** [comment]
I do not know.  The instructions do not link to the bug report, so I cannot check.  `x86_64-pc-linux-gnu-g++-7.1.0` works for me, but that is not the compiler in question (that `g++` generates code for Linux, and the instructions discuss generating code for Windows).  Based on context, that statement probably refers to `i686-w64-mingw32-g++-7`, which I do not use.  I tried installing `i686-w64-mingw32-g++-6` yesterday prior to sending you back for another attempt, but the install failed, so I tested with what I had available (`-5.4.0`) and then pushed the changes for you to try.  Fortunately, that worked for you.  I did not take the time to investigate why `-6` failed for me, nor have I tried `-7` or `-8` at all.

@Mako88: did upstream ever address the gcc-7 Windows problem?  If so, what do users need to check to verify that they get a good copy?

Closing ticket as the original report was about a compilation failure, and OP reports that the program now builds successfully.  The ticket can still be used, but I want it closed so people do not think the build is still broken.
